### DISCUSSION TITLE
3d Thermal conduction problem

### DISCUSSION THREAD

**Discussion by hbadi** [mfem/mfem#3522] [Q&A] [answered]
Hi,

This is my first example with mfem, i got inspired from many mfem examples. 
I wish to simulation a "basic" 3d thermal conduction problem, each group of solids has its own conductivity (marker 0 to 6) and with the two boundary conditions : 
- exterior space (maker=1) : lambda*grad(u).n = (1/R_e)*(u-u_e)
- interior space (maker=2) : lambda*grad(u).n = (1/R_i)*(u-u_i)
- adiabatic space (marker=3) : lambda.grad(u).n = 0
u_i and R_i are respectively the space temperature and the thermal resistance.

Using the dirichlet bc give the good result but when it comes to use Robin bc, the temperature is 0. There's something i can't figure out in the bilinear/Linear forms...

Any idea to fix that problem ?

Moreover, when profiling my code it seems that `a.Assemble` is very time consuming, it there a way to speedup ?

The source code and mesh file can be found in that : [google drive](https://drive.google.com/drive/folders/17_UWxK119pZZz2RlTe7X7dUF_O-NEUKX?usp=sharing)

Thank you for helping !

**Comment by mlstowell** [comment, marked as answer]
Hello, @hbadi,

I believe the problem is that `boundary_dofs` should be empty when applying Robin BCs. Any dofs listed in `boundary_dofs` are assumed to be Dirichlet BCs and so the values in `u` are taken to be the Dirichlet values and these will override any corresponding values in the right hand side vector. Since the only non-zero values in `b` were overridden with the zeros in `u` your solution was zero everywhere.

I think that should help although I was not able to test this (my SuiteSparse installation may be corrupted).

Matrix assembly is often one of the most expensive steps, if not the most expensive step, in a finite element simulation. There may be ways to speed this up for your problem but I'm not the best person to offer such advice. @camierjs or @YohannDudouit could probably offer more useful advice on this topic.

Best wishes,
Mark

**Comment by hbadi** [comment]
@Hi @mlstowell 

- You're right, `boundary_dofs ` must be empty. Thanks !
- `AssemblyLevel::LEGACY` assembly is longer than the solving step, i wish assembly could be faster.
Best regards.

**Reply by v-dobrev** [reply]
Regarding the assembly cost: have you tried `AssemblyLevel::FULL`?

Another option that may work for you is to try the version of `ex1`/`ex1p` from `miniapps/performance`.

**Reply by hbadi** [reply]
Hello @v-dobrev,

I tried `AssemblyLevel::FULL` (and also other mode assembly) but i got the error 
```Verification failed: (mode == DofToQuad::FULL) is false:
 --> invalid mode requested
 ... in function: virtual const mfem::DofToQuad& mfem::ScalarFiniteElement::GetDofToQuad(const mfem::IntegrationRule&, mfem::DofToQuad::Mode) const
 ... in file: C:/M/mingw-w64-mfem/src/mfem-4.5/fem/fe/fe_base.cpp:500```

The miniapp `ex1/ex1p` use a lot of trick that I can't figure one by one what is the most relevant for me.

Best regards.

**Reply by v-dobrev** [reply]
With `Assembly::FULL`, what kind of mesh are you using: quad/hex (tensor product elements) or triangle/tet (simplex elements)? If it is a simplex mesh, that is (generally) not supported yet. If the mesh is quads/hexes that may be an indication of a bug.
